### PR TITLE
Parser of input data from .csv file and it's validation

### PR DESCRIPTION
**Pull request by vanoha**
Now data can be represented as .csv files. New feature reads .csv files and validates them.

### CODE REVIEW

File: sampo/userinput/parser/csv_parser.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        # 2. gather resources and contractors based on work resource estimator or contractor info .csv file
+        contractors = []
+        resource_names = []
+        works_info['activity_name_original'] = works_info.activity_name
+        if unique_work_names_mapper:
+            works_info.activity_name = works_info.activity_name.apply(lambda name: unique_work_names_mapper[name])
+
+        if contractor_info is None and isinstance(work_resource_estimator, DefaultWorkEstimator):
+            raise InputDataException(
+                'you have neither info about contractors nor work resource estimator.'
```
Comment (StannisMod) [inline sampo/userinput/parser/csv_parser.py:63]: `DefaultWorkTimeEstimator` should generate default resources pool

File: sampo/userinput/parser/general_build.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                predecessor_ids[-1].append(row['predecessor_ids'][index])
+                connection_types[-1].append(row['connection_types'][index])
+                lags[-1].append(row['lags'][index])
+        if len(predecessor_ids[-1]) == 0:
+            predecessor_ids[-1].append(NONE_ELEM)
+            connection_types[-1].append(EdgeType.FinishStart)
+            lags[-1].append(float(NONE_ELEM))
+    frame['predecessor_ids'], frame['connection_types'], frame['lags'] = predecessor_ids, connection_types, lags
+
+    frame['edges'] = frame[['predecessor_ids', 'connection_types', 'lags']].apply(lambda row: list(zip(*row)), axis=1)
```
Comment (StannisMod) [inline sampo/userinput/parser/general_build.py:117]: You can merge `115` and `117` strings

File: sampo/userinput/parser/general_build.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    return frame
+
+
+def topsort_graph_df(frame: pd.DataFrame) -> pd.DataFrame:
+    frame['predessors_sh'] = [tuple(zip(*list(zip(*edges))[:2])) for edges in frame['edges']]
```
Comment (StannisMod) [inline sampo/userinput/parser/general_build.py:122]: `predessors_sh` column maybe not used by outer code

File: sampo/userinput/parser/general_build.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    has_succ = set()
+    id_to_node = {NONE_ELEM: start}
+
+    for _, row in frame.iterrows():
+        if 'min_req' in frame.columns and 'max_req' in frame.columns:
+            reqs = [WorkerReq(res_name, row[res_name],
+                              min(row['min_req'][res_name], workers_max_count[res_name]),
+                              min(row['max_req'][res_name], workers_max_count[res_name])
+                              ) for res_name in resource_names
+                    if 0 < row['min_req'][res_name] <= row['max_req'][res_name]]
```
Comment (StannisMod) [inline sampo/userinput/parser/general_build.py:159]: `min_req` can be zero

File: sampo/userinput/parser/history.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        for j in range(i, len(tasks_df)):
+            if i != j:
+                works1_ids.append(tasks_df.loc[i, 'activity_id'])
+                works2_ids.append(tasks_df.loc[j, 'activity_id'])
+                if use_mapper:
+                    works1_names.append(mapper[tasks_df.loc[i, 'granular_name']])
+                    works2_names.append(mapper[tasks_df.loc[j, 'granular_name']])
+                else:
+                    works1_names.append(tasks_df.loc[i, 'granular_name'])
+                    works2_names.append(tasks_df.loc[j, 'granular_name'])
```
Comment (StannisMod) [inline sampo/userinput/parser/history.py:26]: Rewrite without copy-paste (without `else`)

File: sampo/userinput/parser/history.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+    return {"ids": works1_ids, "names": works1_names}, {"ids": works2_ids, "names": works2_names}
+
+
+def get_delta_between_dates(first: str, second: str) -> int:
+    if type(first) == float or type(second) == float:
+        return 0
```
Comment (StannisMod) [inline sampo/userinput/parser/history.py:33]: You are checking input type that does not conforms to type contract

File: sampo/userinput/parser/history.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    return max((datetime.date(int(first.split('-')[0]), int(first.split('-')[1]), int(first.split('-')[2])) -
+                datetime.date(int(second.split('-')[0]), int(second.split('-')[1]), int(second.split('-')[2]))).days, 1)
+
+
+def find_min_without_outliers(lst: list[float]):
```
Comment (StannisMod) [inline sampo/userinput/parser/history.py:38]: Return type?

File: sampo/userinput/parser/history.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+    return round(min([x for x in lst if x >= np.mean(lst) - 3 * np.std(lst)]), 2)
+
+
+# Собираем статистику по встречаемости разного взаимного расположения задач
+# s1, f1 - старт и финиш первой работы, s2 и f2 - второй работы
```
Comment (StannisMod) [inline sampo/userinput/parser/history.py:43]: Oops, Russian comments...

**Review by StannisMod — DISMISSED** [review]
Requested cosmetic changes

**Comment by StannisMod** [comment]
Also, you should add data schema of the `WorkGraph` format

File: sampo/userinput/parser/general_build.py (hunk 2 above)
Comment (vanoha) [reply sampo/userinput/parser/general_build.py]: No, I can't
In the `topsort_graph_df` function I need to have `predecessor_ids` column

**Review by vanoha — COMMENTED** [review]
Changes done

**Review by vanoha — COMMENTED** [review]
Changes done

**Review by angrymuskrat — COMMENTED** [review]
norm

**Comment by angrymuskrat** [comment]
norm

**Review by angrymuskrat — APPROVED** [review]

**Pull request merged**